### DISCUSSION TITLE
GitHub Action runtimes for a larger set of cross-compilers on Ubuntu Focal

### DISCUSSION THREAD

**Discussion by junlarsen** [bytedeco/javacpp-presets#1009] [Ideas]
I've been working on a preset for gcc/libgccjit as you probably know from the mailing lists. To build gcc we require a more modern version of the gcc compiler, that's just how gcc's build works.

I recently came across a package on the launchpad ppa for Ubuntu which seems to have cross compilers for gcc-10 for a lot of the architectures we have presets for.

- https://launchpad.net/ubuntu/+source/gcc-10-cross
- https://packages.ubuntu.com/source/focal/gcc-10-cross

It looks like this package has cross compilers for aarch64, arm-gnueabi, arm-gnueabihf, powerpc, powerpc64le, i686, s390x and x86-64.

Lacking experience with apt and ubuntu's packages I'm assuming the packages published for Focal/20.04 are probably incompatible with Xenial/16.04 which is the version the actions are running on. If this is the case, an option would be to upgrade the actions themselves to 20.04, but I imagine there will be some migration issues so it's probably better to add an Ubuntu Focal action instead.

I'll build an environment to run the actions on for the libgccjit preset, but I think it's something we could look to move into the javacpp-presets repo along with the other actions for building presets. Let me know what you think.

**Comment by saudet** [comment]
I'm pretty sure GCC's build can bootstrap itself: https://gcc.gnu.org/install/build.html
It's probably not something that we have to worry about. Have you tried?

**Reply by junlarsen** [reply]
I see, perhaps I can use the existing gcc compiler to build the cross compiler and the necessary binutils and then use the cross compiler to get libgccjit? I suppose it's the same as using the ppa package except we built the cross compiler ourselves?

Forgot to mention, but my current block on the gcc preset is the older cross compilers not supporting the -fself-test option to test the bootstrapped compiler created during the build which is why I'd like a more recent cross compiler. I already tried messing around in the generated makefiles and such but I didn't have any luck doing that as I ran into more problems further down the line.

**Reply by saudet** [reply]
AFAIU, the bootstrap builds *everything*, including libgccjit. It should work with anything that comes with GCC 4.8 or later installed: https://gcc.gnu.org/install/prerequisites.html

**Reply by junlarsen** [reply]
Cool, seems like this is the case. I think I have a working build for ppc64le now. Will do further testing and research over the weekend and look into the other targets I have in mind.

Does indeed seem like the gcc build script requires a newer cross gcc for the self tests during build even though we're compiling with the host gcc.  There might even be an option to disable the self tests somewhere. Should be able to run everything from the existing build scripts if we just install the newer cross-compilers on the build machine.

Looking good so far, excited to see how this works out.

**Reply by junlarsen** [reply]
After performing a couple experiments it seems the packaged cross-compilers in the default repositories do not have versions which accept the -fself-test flag. I honestly thought gcc-5 which is available would do, but it seems like GCC 5 doesn't understand it either.

It also seems to be the case that building with the host x86-64 gcc compiler produces an x86-64 ELF shared object even though the code should be built for the ppc64le-linux-gnu target. This elf file obviously won't succeed in linking so this route leads nowhere. 

I do however have a fully working build using the gcc-10 cross compilers which runs fine on 20.04. Not sure how easy it'd be to integrate the CI/envs for the gcc preset into the GitHub actions setup we have for the other presets. If we wish to do this we'd need a way to get the newer cross compilers onto the older ubuntu systems (or alternatively look for a different distro/image which has them available). There may be ppa's out there which also host these new compilers, but I didn't find anything useful with a quick search.

Another option would be to build the cross compilers and distribute them ourselves, but I think that's additional work which should be a longer term goal if we decide to head down that route as it's not necessary to get the preset up and going. We should note that custom build envs to grab a newer compiler is a less common case which most presets won't need because gcc requires a recent version of its cross compiler to cross build itself so I think it's fine leaving the CI for this preset by itself.

For now I think I'll just get the builds running on 20.04 because it seems like it's working and then we'll see if we want to migrate the CI into the main repository further down the line.

**Reply by saudet** [reply]
The problem isn't about the version of Ubuntu, the problem is that although it may work now to rely on random builds available out there, there is no guarantees that they will remain available in the future. One thing you could do is look at how those builds are done, and replicate that in our builds. How does that sound?

**Reply by saudet** [reply]
Also, having builds that work on no other version of Linux than the very latest bleeding-edge version of Ubuntu isn't very useful, so it may be best to simply not offer anything until there is actual demand for such builds. Sounds OK?

**Reply by saudet** [reply]
But that shouldn't prevent us from having the repository here. I want to move all presets in their own repositories, eventually, so having GCC there as a separate "experimental" repository that contains its own set of actions that work only with Focal, that would be fine. I didn't realize you were not a member here yet, so I've sent an invitation. You should be able to move your repository here any time you feel like it: https://docs.github.com/en/github/administering-a-repository/transferring-a-repository

**Reply by junlarsen** [reply]
Sounds good. I've added an experimental notice to the build which runs on that environment and moved the repository over.
